### PR TITLE
feat: add accept_application route and usecase

### PR DESCRIPTION
**Pull request by tdelabro**

### CODE REVIEW

File: src/domain/repositories/application.rs [hunk 1]
```diff
@@ -22,4 +22,5 @@ pub trait Repository: Send + Sync {
 		contribution_id: &ContributionId,
 		contributor_id: &Option<ContributorId>,
 	) -> Result<Vec<Application>, Error>;
+	fn accept_application(&self, id: &ApplicationId) -> Result<Application, Error>;
```
Comment (Bernardstanislas) [inline src/domain/repositories/application.rs:25]: This should not live in a repository, this is some business action. This should be in a `ApplicationService`, which does the action and then calls a `save` method on this repository.
Repositories are pure data access services, not business logic holders

File: src/routes/applications/accept/tests.rs [hunk 2]
```diff
@@ -0,0 +1,87 @@
+use super::accept_application;
```
Comment (Bernardstanislas) [inline src/routes/applications/accept/tests.rs:1]: ❤️

**Review by Bernardstanislas — CHANGES_REQUESTED** [review]

File: src/domain/repositories/application.rs (hunk 1 above)
Comment (tdelabro) [reply src/domain/repositories/application.rs]: done

File: src/infrastructure/database/tests/application_repository.rs [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+	assert_eq!(found_application, Some(application2));
+
+	let found_application =
+		<Client as ApplicationRepository>::find(&client, &Uuid::new_v4().into()).unwrap();
+	assert_eq!(found_application, None);
+}
+
+#[test]
+#[ignore = "require a database"]
+fn id_must_be_unique() {
```
Comment (tekkac) [inline src/infrastructure/database/tests/application_repository.rs:49]: I am not sure but I recall we discussed that `contributor_id` can only apply once, but not sure if it is specified in this sprint or User Story

File: src/routes/applications/accept/mod.rs [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+	usecase: &State<Box<dyn AcceptApplicationUsecase>>,
+) -> Result<status::Accepted<()>, HttpApiProblem> {
+	let application_id: ApplicationId = (*application_id.as_uuid()).into();
+
+	usecase
+		.accept_application(&application_id)
+		.map_err(|e| e.to_http_api_problem())?;
+
+	// TODO after action queue is removed:
+	// return the hash of the on-chain transaction containing the call
```
Comment (tekkac) [inline src/routes/applications/accept/mod.rs:27]: I assume this is for another US?

File: src/infrastructure/database/domain_implementations/application/service.rs [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+					applications::dsl::applications.filter(applications::id.eq(id.as_uuid())),
+				)
+				.set(applications::status.eq(Status::Accepted))
+				.get_result::<models::Application>(&*connection)
+				.map_into()?;
+
+				// Set all other pending applications to refused
+				diesel::update(
+					applications::dsl::applications
+						.filter(applications::status.eq(Status::Pending)),
```
[the remark this answers is not in the document]
Comment (tdelabro) [reply src/infrastructure/database/domain_implementations/application/service.rs]: done

File: src/bin/utils/mod.rs [hunk 6]
```diff
@@ -46,18 +46,18 @@ pub fn find_project_by_title(projects: &Value, title: &'static str) -> serde_jso
...
+		.unwrap_or_else(|| panic!("could not find {title}"))
 		.to_owned()
 }
 
 pub fn find_contribution_by_onchain_id(project: &Value, contribution_onchain_id: u64) -> Value {
 	project["contributions"]
 		.as_array()
 		.expect("contributions is not an array")
-		.into_iter()
-		.find(|contribution| contribution["onchain_id"] == contribution_onchain_id.to_string())
```
[the remark this answers is not in the document]
Comment (tdelabro) [reply src/bin/utils/mod.rs:60]: fixed

File: src/infrastructure/database/domain_implementations/application/repository.rs [hunk 7]
```diff
@@ -82,12 +74,14 @@ impl From<models::Application> for Application {
...
 			application.contribution_id.into(),
 			ContributorId::from(application.contributor_id.as_str()),
+			application.status.into(),
 		)
 	}
 }
 
 impl From<diesel::result::Error> for ApplicationRepositoryError {
 	fn from(error: diesel::result::Error) -> Self {
+		println!("kind: {:?}", error);
```
[the remark this answers is not in the document]
Comment (tdelabro) [reply src/infrastructure/database/domain_implementations/application/repository.rs:84]: done

File: src/infrastructure/database/tests/application_repository.rs (hunk 3 above)
Comment (tdelabro) [reply src/infrastructure/database/tests/application_repository.rs:49]: tested in `cannot_apply_twice`

File: src/routes/applications/accept/mod.rs (hunk 4 above)
Comment (tdelabro) [reply src/routes/applications/accept/mod.rs:27]: yup

File: src/infrastructure/database/domain_implementations/application/service.rs [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+impl From<diesel::result::Error> for ApplicationServiceError {
+	fn from(error: diesel::result::Error) -> Self {
+		match error {
+			diesel::result::Error::DatabaseError(kind, _) => match kind {
+				diesel::result::DatabaseErrorKind::UniqueViolation =>
+					Self::AlreadyExist(Box::new(error)),
+				diesel::result::DatabaseErrorKind::ForeignKeyViolation =>
+					Self::InvalidEntity(Box::new(error)),
```
[the remark this answers is not in the document]
Comment (tdelabro) [reply src/infrastructure/database/domain_implementations/application/service.rs:52]: thanks!

**Pull request merged**